[flutter_tools] Auto-detection of unsound null safety should fail for web dev builds

Any application that requires unsound null safety should be forced to pass an explicit flag when running `flutter run -d chrome`. Auto-detecting unsound null safety should be an error.

It sounds like this is working as intended for all implementations except development mode for the web (JS compiler).

I reproduction case should be to create an app and add a language version comment to the library that contains the entrypoint:

```
// @dart = 2.11
```

Then run `flutter run -d chrome`. It prints a warning but that should be an error. 
Running with the explicit flag: `flutter run -d chrome --no-sound-null-safety` should still work with a warning.

Reproducible using the code sample below. 

<details>
<summary>code sample</summary>

```dart
// @dart = 2.11
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'ReorderableListView + Dismissible Items'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key key, @required this.title}) : super(key: key);

  final String title;
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: const Center(
        child: Text('some text'),
      ),
    );
  }
}

```
</details>

<details>
<summary>logs</summary>

```
Launching lib/main.dart on Chrome in debug mode...
This app is linked to the debug service: ws://127.0.0.1:50964/TFe4lan2NJg=/ws
Debug service listening on ws://127.0.0.1:50964/TFe4lan2NJg=/ws
Running without sound null safety ⚠️
Dart 3 will only support sound null safety, see https://dart.dev/null-safety
Connecting to VM Service at ws://127.0.0.1:50964/TFe4lan2NJg=/ws
```

```
Launching lib/main.dart on macOS in debug mode...
--- xcodebuild: WARNING: Using the first of multiple matching destinations:
{ platform:macOS, arch:arm64, id:00008103-000329201AD1001E }
{ platform:macOS, arch:x86_64, id:00008103-000329201AD1001E }
: Error: Library doesn't support null safety.
// @dart = 2.11
^^^^^^^^^^^^^^^

Target kernel_snapshot failed: Exception

Command PhaseScriptExecution failed with a nonzero exit code
** BUILD FAILED **

Exception: Build process failed
Exited
```

```
Launching lib/main.dart on iPhone 14 in debug mode...
Updating project for Xcode compatibility.
Upgrading project.pbxproj
Upgrading Runner.xcscheme
Removing script build phase dependency analysis.
Xcode build done.                                           34.7s
Failed to build iOS app
Error (Xcode): lib/main.dart:1:1: Error: Library doesn't support null safety.
Could not build the application for the simulator.
Error launching application on iPhone 14.
Exited
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[!] Flutter (Channel master, 3.7.0-19.0.pre.13, on macOS 13.1 22C65 darwin-arm64, locale en-GB)
    • Flutter version 3.7.0-19.0.pre.13 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter
      SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK
      checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f8628b5cb1 (4 hours ago), 2023-01-11 18:48:03 -0800
    • Engine revision 4a8d6866a1
    • Dart version 3.0.0 (build 3.0.0-112.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform
      update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.74.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 109.0.5414.87

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```
</details>

--- a/packages/flutter_tools/lib/src/resident_runner.dart
+++ b/packages/flutter_tools/lib/src/resident_runner.dart
@@ -83,10 +83,8 @@ class FlutterDevice {
     required Platform platform,
     TargetModel targetModel = TargetModel.flutter,
     List<String>? experimentalFlags,
-    ResidentCompiler? generator,
     String? userIdentifier,
   }) async {
-    ResidentCompiler generator;
     final TargetPlatform targetPlatform = await device.targetPlatform;
     if (device.platformType == PlatformType.fuchsia) {
       targetModel = TargetModel.flutterRunner;
@@ -111,6 +109,8 @@ class FlutterDevice {
       fileSystem: globals.fs,
     );
 
+    final ResidentCompiler generator;
+
     // For both web and non-web platforms we initialize dill to/from
     // a shared location for faster bootstrapping. If the compiler fails
     // due to a kernel target or version mismatch, no error is reported
@@ -119,7 +119,7 @@ class FlutterDevice {
     // used to file a bug, but the compiler will still start up correctly.
     if (targetPlatform == TargetPlatform.web_javascript) {
       // TODO(zanderso): consistently provide these flags across platforms.
-      late String platformDillName;
+      final String platformDillName;
       final List<String> extraFrontEndOptions = List<String>.of(buildInfo.extraFrontEndOptions);
       if (buildInfo.nullSafetyMode == NullSafetyMode.unsound) {
         platformDillName = 'ddc_outline.dill';
@@ -132,12 +132,12 @@ class FlutterDevice {
           extraFrontEndOptions.add('--sound-null-safety');
         }
       } else {
-        assert(false);
+        throw StateError('Expected buildInfo.nullSafetyMode to be one of unsound or sound, got ${buildInfo.nullSafetyMode}');
       }
 
       final String platformDillPath = globals.fs.path.join(
         getWebPlatformBinariesDirectory(globals.artifacts!, buildInfo.webRenderer).path,
-        platformDillName
+        platformDillName,
       );
 
       generator = ResidentCompiler(

--- a/packages/flutter_tools/lib/src/runner/flutter_command.dart
+++ b/packages/flutter_tools/lib/src/runner/flutter_command.dart
@@ -1093,7 +1093,11 @@ abstract class FlutterCommand extends Command<void> {
             (languageVersion.major == nullSafeVersion.major && languageVersion.minor >= nullSafeVersion.minor)) {
           nullSafetyMode = NullSafetyMode.sound;
         } else {
-          nullSafetyMode = NullSafetyMode.unsound;
+          throwToolExit(
+            'This application does not support sound null-safety (its language version is $languageVersion).\n'
+            'To build this application, you must provide the CLI flag --no-sound-null-safety. Dart 3 will only '
+            'support sound null safety, see https://dart.dev/null-safety.',
+          );
         }
       } else if (!wasNullSafetyFlagParsed) {
         // This mode is only used for commands which do not build a single target like

--- a/packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart
+++ b/packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart
@@ -554,6 +554,27 @@ void main() {
       ProcessManager: () => processManager,
     });
 
+    testUsingContext('tool exits on non-sound-null-safe code when explicit flag not passed', () async {
+      final DummyFlutterCommand flutterCommand = DummyFlutterCommand(packagesPath: 'foo');
+      flutterCommand.argParser
+          ..addFlag(FlutterOptions.kNullSafety, defaultsTo: true)
+          ..addOption('target');
+      final File targetFile = fileSystem.file('targetFile.dart')
+          ..writeAsStringSync('// @dart = 2.11');
+      expect(
+        () async => flutterCommand.getBuildInfo(
+          forcedBuildMode: BuildMode.debug,
+          forcedTargetFile: targetFile,
+        ),
+        throwsToolExit(
+          message: 'This application does not support sound null-safety (its language version is 2.11)',
+        ),
+      );
+    }, overrides: <Type, Generator>{
+      FileSystem: () => fileSystem,
+      ProcessManager: () => processManager,
+    });
+
     testUsingContext('use packagesPath to generate BuildInfo', () async {
       final DummyFlutterCommand flutterCommand = DummyFlutterCommand(packagesPath: 'foo');
       final BuildInfo buildInfo = await flutterCommand.getBuildInfo(forcedBuildMode: BuildMode.debug);

--- a/packages/flutter_tools/test/integration.shard/test_data/stepping_project.dart
+++ b/packages/flutter_tools/test/integration.shard/test_data/stepping_project.dart
@@ -65,7 +65,7 @@ class WebSteppingProject extends Project {
   final String pubspec = '''
   name: test
   environment:
-    sdk: '>=2.10.0 <4.0.0'
+    sdk: '>=2.12.0 <4.0.0'
   dependencies:
     flutter:
       sdk: flutter

